### PR TITLE
Modify plan-traversal algorithm

### PR DESCRIPTION
**Pull request by wolffcm**
Also:
- Added a new `ID()` method to `PlanNode`
- Added ability to add rules that match any node
- Added plan traversal test

### CODE REVIEW

File: planner/plan_traversal_test.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
+package planner
```
Comment (jlapacik) [inline planner/plan_traversal_test.go]: Convention is that tests are placed in a separate package. Is there a reason for having them in the same package?

File: planner/planner.go [hunk 2]
```diff
@@ -5,102 +5,122 @@ type Planner interface {
...
+				newNode, changed := p.matchRules(node)
+				anyChanged = anyChanged || changed
+				if node != newNode {
+					updateSuccessors(inputPlan, node, newNode)
+				}
+
+				// append to stack in reverse order so lower-indexed children
+				// are visited first.
+				for i := len(newNode.Predecessors()); i > 0; i-- {
+					nodeStack = append(nodeStack, newNode.Predecessors()[i - 1])
```
Comment (jlapacik) [inline planner/planner.go]: What about successors? If a node (call it Node) is rewritten, there could conceivably be a rule that previously did not match a successor of Node but now does. Hence that successor should be visited again, right?

File: planner/planner.go [hunk 3]
```diff
@@ -5,102 +5,122 @@ type Planner interface {
...
+// Plan may change its argument and/or return a new instance of QueryPlan, so the correct way to call Plan is:
+//     plan, err = planner.Plan(plan)
+func (p *LogicalToPhysicalPlanner) Plan(inputPlan *QueryPlan) (*QueryPlan, error) {
+
+	visited := make(map[PlanNode]struct{})
+
+	nodeStack := make([]PlanNode, len(inputPlan.Roots()))
+	copy(nodeStack, inputPlan.Roots())
+
+	for anyChanged := true; anyChanged == true; {
```
Comment (jlapacik) [inline planner/planner.go]: Nice.

File: planner/plan_traversal_test.go (hunk 1 above)
Comment (wolffcm) [reply planner/plan_traversal_test.go]: No, no reason.  I did not realize that it was a convention.  I'll change it.

**Pull request merged**